File: test/sparkmax-test/src/main/cpp/Robot.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "Robot.h"

#include <frc/smartdashboard/SmartDashboard.h>
#include <frc2/command/CommandScheduler.h>
#include <frc/XboxController.h>
#include <frc/drive/DifferentialDrive.h>
#include <iostream>
#include <frc/smartdashboard/SmartDashboard.h>

void Robot::RobotInit() {
}

/**
 * This function is called every robot packet, n>o matter the mode. Use
 * this for items like diagnostics that you want to run during disabled,
 * autonomous, teleoperated and test.
 *
 * <p> This runs after the mode specific periodic functions, but before
 * LiveWindow and SmartDashboard integrated updating.
 */
void Robot::RobotPeriodic() {
  frc2::CommandScheduler::GetInstance().Run();
}

/**
 * This function is called once each time the robot enters Disabled mode. You
 * can use it to reset any subsystem information you want to clear when the
 * robot is disabled.
 */
void Robot::DisabledInit() {}

void Robot::DisabledPeriodic() {}

/**
 * This autonomous runs the autonomous command selected by your {@link
 * RobotContainer} class.
 */
void Robot::AutonomousInit() {
  m_autonomousCommand = m_container.GetAutonomousCommand();

  if (m_autonomousCommand != nullptr) {
    m_autonomousCommand->Schedule();
  }

  m_motorStart = std::chrono::system_clock::now();
  m_motor.Set(0.0);
}

void Robot::AutonomousPeriodic() {
  std::chrono::time_point now = std::chrono::system_clock::now();
  std::chrono::duration<double> timeSinceStart = now - m_motorStart;
  double power = timeSinceStart.count() * 0.2;  
  m_motor.Set((power) > MAX_POWER ? MAX_POWER : (power)); 
  std::cout << m_motor.Get() << '\n';
  frc::SmartDashboard::PutNumber("spark max power", m_motor.Get());
}

void Robot::TeleopInit() {
  m_motor.Set(0.0);
}

/**
 * This function is called periodically during operator control.
 */
void Robot::TeleopPeriodic() { 
  // This makes sure that the autonomous stops running when
  // teleop starts running. If you want the autonomous to
  // continue until interrupted by another command, remove
  // this line or comment it out.
  if (m_autonomousCommand != nullptr) {
    m_autonomousCommand->Cancel();
    m_autonomousCommand = nullptr;
  }

  static bool pressed = false;

  //If A is pressed, start spinning the motor
  if(m_xbox.GetAButton())
  {
    if(!pressed)
    {
      m_motorStart = std::chrono::system_clock::now();
      pressed = true;
    }
    std::chrono::time_point<std::chrono::system_clock> now = std::chrono::system_clock::now();
    std::chrono::duration<double> timeSinceStart = now - m_motorStart; 
    double power = timeSinceStart.count() * 0.05;
    if(power > MAX_POWER)
      power = MAX_POWER; 
    m_motor.Set(power);
    frc::SmartDashboard::PutNumber("spark max power", m_motor.Get());
    std::cout << m_motor.Get() << '\n';
  }
  else
  {
    static std::chrono::time_point<std::chrono::system_clock> 
      start = std::chrono::system_clock::now(), 
      end = std::chrono::system_clock::now();

    if(pressed)
      start = std::chrono::system_clock::now();

    end = std::chrono::system_clock::now();
  
    double currentPower = m_motor.Get();
    //Slowly turn off the robot
    if(currentPower > 0.0)
    {
      std::chrono::duration<double> timePassed = end - start;
      currentPower -= 0.5 * timePassed.count();
    }
    else if(currentPower < 0.0)
      currentPower = 0.0;
    m_motor.Set(currentPower);

    pressed = false;
    frc::SmartDashboard::PutNumber("spark max power", m_motor.Get());
    std::cout << m_motor.Get() << '\n';

    start = std::chrono::system_clock::now(); 
  }
}

/**
 * This function is called periodically during test mode.
 */
void Robot::TestPeriodic() {
  
}

#ifndef RUNNING_FRC_TESTS
int main() {
  return frc::StartRobot<Robot>();
}
#endif

File: test/swervebase/src/main/java/frc/robot/Robot.java
package frc.robot;

import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.kinematics.SwerveModuleState;
import edu.wpi.first.networktables.GenericEntry;
import edu.wpi.first.networktables.NetworkTableEntry;
import edu.wpi.first.wpilibj.TimedRobot;
import edu.wpi.first.wpilibj.shuffleboard.Shuffleboard;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj2.command.Command;
import edu.wpi.first.wpilibj2.command.CommandScheduler;
import frc.robot.subsystems.SwerveModule;

/**
 * The VM is configured to automatically run this class, and to call the functions corresponding to
 * each mode, as described in the TimedRobot documentation. If you change the name of this class or
 * the package after creating this project, you must also update the build.gradle file in the
 * project.
 */
public class Robot extends TimedRobot {

  SwerveModule mod;
  GenericEntry off;
    GenericEntry vel;

      GenericEntry inpPos;

  /**
   * This function is run when the robot is first started up and should be used for any
   * initialization code.
   */
  @Override
  public void robotInit() {
    double THRUST_DISTANCE_PER_TICK = .0254 * 4.0 * Math.PI / (2048.0 * 6.75);
    double THRUST_KV = 1023.0 * 600.0 / (6380.0 * 2048.0);
    double STEER_KV = 1.0/300.0;
    mod=new SwerveModule(new SwerveModuleConstants(
            STEER_KV, 0, 0, 0, 0.067, THRUST_KV, 0, 0, 0, THRUST_DISTANCE_PER_TICK, 31, 1));
    // Instantiate our RobotContainer.  This will perform all our button bindings, and put our
    // autonomous chooser on the dashboard.


    off=Shuffleboard.getTab("TAASAS").add("Offset", mod.constants.steerOffsetRotations).getEntry();
    vel=Shuffleboard.getTab("TAASAS").add("vel", 0.0).getEntry();

    inpPos=Shuffleboard.getTab("TAASAS").add("inppos", 0.0).getEntry();


  }

  /**
   * This function is called every robot packet, no matter the mode. Use this for items like
   * diagnostics that you want ran during disabled, autonomous, teleoperated and test.
   *
   * <p>This runs after the mode specific periodic functions, but before LiveWindow and
   * SmartDashboard integrated updating.
   */
  @Override
  public void robotPeriodic() {
    // Runs the Scheduler.  This is responsible for polling buttons, adding newly-scheduled
    // commands, running already-scheduled commands, removing finished or interrupted commands,
    // and running subsystem periodic() methods.  This must be called from the robot's periodic
    // block in order for anything in the Command-based framework to work.

    mod.constants.steerOffsetRotations=off.getDouble(mod.constants.steerOffsetRotations);
    //SmartDashboard.putNumber("Pos",mod.constants.steerOffsetRotations+mod.getModulePosition().angle.getRotations());

  }

  /** This function is called once each time the robot enters Disabled mode. */
  @Override
  public void disabledInit() {}

  @Override
  public void disabledPeriodic() {}

  /** This autonomous runs the autonomous command selected by your {@link RobotContainer} class. */
  @Override
  public void autonomousInit() {

  }

  /** This function is called periodically during autonomous. */
  @Override
  public void autonomousPeriodic() {}

  @Override
  public void teleopInit() {

    // This makes sure that the autonomous stops running when
    // teleop starts running. If you want the autonomous to
    // continue until interrupted by another command, remove
  }

  /** This function is called periodically during operator control. */
  @Override
  public void teleopPeriodic() {
        double v=vel.getFloat((float) -0.87);
    double r=inpPos.getFloat(0);
    mod.setModuleState(new SwerveModuleState(v,Rotation2d.fromRotations(r)));

  }

  @Override
  public void testInit() {
    // Cancels all running commands at the start of test mode.
  }

  /** This function is called periodically during test mode. */
  @Override
  public void testPeriodic() {}

  /** This function is called once when the robot is first started up. */
  @Override
  public void simulationInit() {}

  /** This function is called periodically whilst in simulation. */
  @Override
  public void simulationPeriodic() {}
}
